### ISSUE TITLE
Auton Changes

### ISSUE THREAD

**Issue by fraktality** [Dreadbot/Recycle-Rush#7]
First off, I'd like to give Mr. King huge props for making 3TA almost work.  Now we need to decide how best to shave off that last 0.1s.  I was looking over the video Mr. Corum posted today and saw the tote being knocked about for about a fourth of a second second before falling into the intake arms--we can decide how we want to go about fixing that today.

Second off (and the reason this is an issue instead of an email):  We should migrate over to using the GlobalHardware model for any future additions and changes to auton.  As it stands, we have to either store a pointer to any hardware components we use in each state class, or access them through a long-winded method of XMLInput.  Having the hardware components stored once as static class members for everyone to see and use avoids this tomfoolery.
en.wikipedia.org/wiki/Don't_repeat_yourself

**Comment by C7C8** [comment]
It's not quite repeating yourself so much as allowing configuration to change at any time. The system is designed with changing configurations in mind - if we switch to static hardware, we lose that ability to change configuration at any random time (even by XML file, if we ever wanted to). Even if we DO go with a static class method, we're still accessing the members of that class instead of the XMLInput class, which isn't really any change (especially since hardware components already are available for everyone to see, in the Config.h file). The simplest change is change from doing this:

XMLInput::getInstance()->getPGroup("some name here");
XMLInput::getInstance()->getMGroup("another name here");

to doing this:

XMLInput\* input = XMLInput::getInstance();
input->getPGroup("some name here");
input->getMGroup("another name here");

...which honestly isn't that difficult or repeating. We already DO have global hardware - the pointer method of storing a pointer is pretty much the same as what you have . If we do switch over to GlobalHardware, we're just going from one model of doing it to almost the exact same model of doing it without any end **behavioral** changes to the robot (just thought of this: by having GlobalHardware you're having two global hardware systems instead of one, which is very redundant). I'd rather focus our time on fixing 3TA and getting PID to a working state than on converting our code (and thus testing it again) to something that offers no real benefit to the tasks at hand.

**Comment by SteveKing** [comment]
On 2015-04-14 08:14 , Parker Stebbins wrote:

> Second off (and the reason this is an issue instead of an email): We 
> should migrate over to using the GlobalHardware model for any future 
> additions and changes to auton. As it stands, we have to either store 
> a pointer to any hardware components we use in each state class, or 
> access them through a long-winded method of XMLInput. Having the 
> hardware components stored once as static class members for everyone 
> to see and use avoids this tomfoolery.

Agreed that this is a good idea, just a word of warning: we may not have 
time to test it properly.  Ugly code that works beats elegant code that 
doesn't. Right now everything but 3TA works just fine.  I strongly 
recommend an "if it ain't broke don't fix it" strategy.

On a related note, one of the main reasons I gave up on 3TA at States 
(besides long lines at the practice area) is that it's tightly coupled 
to 2TA.  As Chris pointed out, 2TA is just a modification of 3TA.  The 
trouble is, any changes to one affects the other. Worse, some of the 
code and constants are used in the other modes.  I changed timing in 2TA 
which wound up breaking the timing in Just Drive mode.

The way I recommend tweaking 3TA at this point is to leave everything 
as-is for all modes except 3TA.  Don't touch anything. Copy all the 
classes, variables, etc. that control 3TA to new names so they can be 
worked on totally independently of anything else. The idea is that the 
other modes work and we don't want to break them. Make it so that no 
changes to 3TA have any effect at all on the other modes.

It's ugly. It's inelegant. If we went back to the beginning of the 
season it's not the way I'd do it.  But given the time and testing 
constraints I think it's the best option at this point.

(And from the practice-what-you-preach department...  Yes, I'm still 
working on totally decoupling all the auton modes, which is a huge 
change to the code structure.  I'm doing it as a learning exercise 
though. I don't expect to ever use it in competition, there's no time to 
test it.)

tl;dr: EVERYTHING BUT 3TA WORKS. MAKE 3TA WORK WITHOUT TOUCHING ANYTHING 
ELSE.  Thank you. :-)

**Comment by C7C8** [comment]
> It's ugly. It's inelegant. If we went back to the beginning of the season it's not the way I'd do it. 

I complete agree with that. Auton is full of crappy workarounds I came up with during testing so I could get to robot to do what I wanted even though the code wasn't what I wanted. I often had to make a lot of quick changes to keep testing, so there are a lot of things in it that make me cringe.

**Comment by SteveKing** [comment]
On 2015-04-14 12:16 , Sourec wrote:

> ```
> It's ugly. It's inelegant. If we went back to the beginning of the
> season it's not the way I'd do it.
> ```
> 
> I complete agree with that. Auton is full of crappy workarounds I came 
> up with during testing so I could get to robot to do what I wanted 
> even though the code wasn't what I wanted. I often had to make a lot 
> of quick changes to keep testing, so there are a lot of things in it 
> that make me cringe.

Don't worry too much about it.  The most common thing heard by 
programmers looking at their own code is, "Who wrote this crap?" :-)

**Comment by C7C8** [comment]
I have fixed the vast majority of the crappy workarounds and hardware-getting issues, and have even fixed a bit of a semi-circular dependancy issue.
- States now have their very own file. No more scrolling through hundreds upon hundreds of lines of code to find the right function. This also means that no two states need to really know about each other, and you don't have to recompile all of autonomous to make a single tweak.
- States now inherit a RoboState abstract base class (finally found the name for it!) that has static members set aside for hardware and tote count values. This means we can now access any hardware from inside the robot without having to provide everything hardware access at once. It also means that there's now more silliness with static functions like HALBot::getToteCount(), which was another awful kludge.
- State return values are now defined in RoboState - defining them as part of HALBot was just silly.

No functionality outside of autonomous was changed at all, nor was any autonomous behavior changed (except that we no longer use XMLInput function calls within autonomous to get hardware because of the aforementioned static hardware thingy)
